### ISSUE TITLE
NotebookCellMetadata conflicts with user defined metadata

### ISSUE THREAD

**Issue by DonJayamanne** [microsoft/vscode#123235]
Based on the type definitions, Notebook Cell metadata has the following properties:
* inputCollapsed
* outputCollapsed
* <anything else> 

However VS Code adds its own custom propperties into this object which causes issues.
<img width="265" alt="Screen Shot 2021-05-06 at 21 10 29" src="https://user-images.githubusercontent.com/1948812/117396767-82173180-aeaf-11eb-841f-289bda726aae.png">

I think these internal properties `lastRunSuccess` `executionOrder` should not be stored in here, as these are available via the `lastExecutionSummary`.

I can see plenty of issues here, its possible for users to edit the metadata and cause issues in VS Code or VS Code blowing away user defined metadata & overwriting it with its own internal state.

@jrieken @roblourens /cc

**Comment by jrieken** [comment]
> I think these internal properties lastRunSuccess executionOrder should not be stored in here, as these are available via the lastExecutionSummary.

We talked about storing execution state not as internal metadata but outside the document. Tho unsure what todays plans are

**Comment by jrieken** [comment]
Tho, generally these "conflicts" can arise we recommend that extension have all their metadata under one key, like `metadata.jupyter.foo` and `metadata.jupyter.bar`

**Comment by DonJayamanne** [comment]
> We recommend that extension have all their metadata under one key, like metadata.jupyter.foo and metadata.jupyter.bar

Then I would suggest bringing back the `custom` property, else its really pointless having an object with ability to store any data, that makes it look like anything can be stored (when your recommendation is the opposite). I.e. someone creating an extension tomorrow might not come across this issue. I.e. i'd prefer if we went back to the older api:

```typescript
    export class NotebookCellMetadata {
        /**
         * Whether a code cell's editor is collapsed
         */
        readonly inputCollapsed?: boolean;

        /**
         * Whether a code cell's outputs are collapsed
         */
        readonly outputCollapsed?: boolean;

        /**
         * Additional attributes of a cell metadata.
         */
        readonly [key: string]: any;

```

Instead it should be:
```typescript
    export class NotebookCellMetadata {
        /**
         * Whether a code cell's editor is collapsed
         */
        readonly inputCollapsed?: boolean;

        /**
         * Whether a code cell's outputs are collapsed
         */
        readonly outputCollapsed?: boolean;

        /**
         * Additional attributes of a cell metadata.
         */
        readonly custom:  Record<string, any>;
```

This way there is never any ambiguity nor any accidental conflicts.

**Comment by jrieken** [comment]
> This way there is never any ambiguity nor any accidental conflicts.

Really? What about the other extension that also stores stuff inside `custom`? You have the exact same problem except for the potential conflict with VS Code's properties. Even with the custom-property extensions are expected to create a "root node" under which their data goes. 

Also note that our strategy is very similar to nbformat: https://nbformat.readthedocs.io/en/latest/format_description.html#metadata

> Because it is a shared namespace, any custom metadata should use a sufficiently unique namespace, such as metadata.kaylees_md.foo = “bar”.

**Comment by DonJayamanne** [comment]
Let me out this another way, 

I'm a new extension author and I choose 
`xyz` as the key (property) and the user files a bug, then we find out there another extension overwriting out data. Now we can tell the user to disable the other extension or we can work with the other extension author to use another key so we don't step over each other. Things are back to normal and things don't break. And we document what keys at user, like Jupyter.

Now let's take the other scenario, me as a new extension author uses
  key as  `extensionorder`, yet nothing works, infact vscode breaks (execution order is not displayed) and there is data loss because vscode is blowing away extension authors data. 
Will vscode change the property, no. Hence the extension author now changes property to `lastrunsuccess`, again there are problems, then again they change to `last runtime`... And on it goes to they think they got it right... It's all a guess work.
Basically extension authors are told you can store anything, but that's not true ,there is absolutely no guarantee that VSCode won't blow it away .
That seems like a poor api and dev experience.

At a minimum vscode should document reserved words/keys. That's the whole point of the documentation (d.ts). 
Or just tell extension authors store it in `custom`.

**Comment by jrieken** [comment]
> key as extensionorder, yet nothing works, infact vscode breaks (execution order is not displayed) and there is data loss because vscode is blowing away extension authors data.

I was not saying that we don't fix storing more metadata than spec'd. The goal is that VS Code stores exactly what it defines in `vscode.d.ts`. 

What I am saying is that I don't buy the argument of telling extension authors to store stuff in `custom` because it has the same potential for conflicts. So, `metadata.xyz` is as wrong as `metadata.custom.xyz`. It should be `metadata.myExtensionId.xyz`

**Comment by DonJayamanne** [comment]
> . It should be metadata.myExtensionId.xyz

Sure anything, as long as it's documented in d.ts. 
TLDR vscode should not overwrite user data when it's avoidable. 

Anyways, to be clear we (Jupyter) are not blocked by this. I filled this when working on some other personal extension.

**Comment by roblourens** [comment]
I can do the work to get execution stuff out of metadata, I was still trying to decide between keeping the execution stuff inside the document internally but in its own object, or keeping it entirely separate. It seems like it's going to be hard to keep it separate. In the extension host they get mixed together again anyway as `latestExecutionSummary`. And why else would the cell view have a progress bar or a single-item toolbar on the left... So I think I want to keep it in a separate object in the document.

**Comment by jrieken** [comment]
Yeah, the most important parts is that execution gets separated and encapsulated into its own thing. Where it ultimately lives isn't the biggest concern (it should be movable with ease) but I don't buy the argument of `latestExecutionSummary`. That and output are the results of an execution whereas execution itself is just something that's transiently there.
